core/job/service/plugin_service.go

```diff
@@ -65,146 +69,93 @@ func (p JobPluginService) Info(_ context.Context, taskName job.TaskName) (*plugi
-		compiler.From(tnnt.GetConfigs()).WithName("proj").WithKeyPrefix(projectConfigPrefix),
-		compiler.From(tnnt.SecretsMap()).WithName("secret"),
-	)
-
-	var pluginConfigs plugin.Configs
-	for key, val := range configs {
-		compiledConf, err := p.engine.CompileString(val, tmplCtx)
-		if err != nil {
-			p.logger.Warn("template compilation encountered suppressed error: %s", err.Error())
-			compiledConf = val
-		}
-		pluginConfigs = append(pluginConfigs, plugin.Config{
-			Name:  key,
-			Value: compiledConf,
-		})
+	// flatten
+	resourceURNs := []job.ResourceURN{}
+	for _, r := range resource.Resources(resources).GetFlattened() {
+		resourceURNs = append(resourceURNs, job.ResourceURN(r.URN()))
 	}
-	return pluginConfigs
+	return resourceURNs, nil
 }
 
-func (p JobPluginService) compileAsset(ctx context.Context, taskPlugin *plugin.Plugin, spec *job.Spec, w window.Window, scheduledAt time.Time) (map[string]string, error) {
-	var jobDestination string
-	if taskPlugin.DependencyMod != nil {
-		var assets map[string]string
-		if spec.Asset() != nil {
-			assets = spec.Asset()
-		}
-		jobDestinationResponse, err := taskPlugin.DependencyMod.GenerateDestination(ctx, plugin.GenerateDestinationRequest{
-			Config: plugin.ConfigsFromMap(spec.Task().Config()),
-			Assets: plugin.AssetsFromMap(assets),
-			Options: plugin.Options{
-				DryRun: true,
-			},
-		})
-		if err != nil {
-			p.logger.Error("error generating destination: %s", err)
-			return nil, err
-		}
-		jobDestination = jobDestinationResponse.Destination
-	}
-
-	interval, err := w.GetInterval(scheduledAt)
+func (p JobPluginService) generateResources(ctx context.Context, rawResource string, visited map[job.ResourceURN][]*resource.Resource, paths map[job.ResourceURN]bool) ([]*resource.Resource, error) {
+	errs := errors.NewMultiError("generate resources")
+	resourceURNs := p.parserFunc(rawResource)
+	resources := []*resource.Resource{}
+	urnToRawResource, err := p._extractorFunc(ctx, p.logger, resourceURNs)
 	if err != nil {
-		return nil, err
+		p.logger.Error(fmt.Sprintf("error when extract ddl resource: %s", err.Error()))
+		return resources, nil
 	}
 
-	var assets map[string]string
-	if spec.Asset() != nil {
-		assets = spec.Asset()
-	}
+	for _, resourceURN := range resourceURNs {
+		resource := &resource.Resource{}
+		resource.UpdateURN(resourceURN.String())
 
-	templates, err := p.engine.Compile(assets, map[string]interface{}{
-		configKeyDstart:        interval.Start.Format(TimeISOFormat),
-		configKeyDend:          interval.End.Format(TimeISOFormat),
-		configKeyExecutionTime: scheduledAt.Format(TimeISOFormat),
-		configKeyDestination:   jobDestination,
-	})
-	if err != nil {
-		p.logger.Error("error compiling asset: %s", err)
-		return nil, fmt.Errorf("failed to compile templates: %w", err)
-	}
+		if paths[resourceURN] {
+			errs.Append(fmt.Errorf("circular reference is detected"))
+			continue
+		}
 
-	return templates, nil
-}
+		if visited[resourceURN] == nil {
```

Comment:
Can you check if the lookup was success or not instead of nil, `if _, ok :=  visited[resourceURN]; !ok {`.
That way it will be more clear if the lookup failed.